From 38136c6a8aa0c8d31a4dbf6f21efc432c14b3506 Mon Sep 17 00:00:00 2001
From: Michael Chang <mchang@suse.com>
Date: Tue, 16 Jun 2026 11:33:23 +0800
Subject: [PATCH 3/4] fs/btrfs: add btrfs-list-subvols command

Add a new "btrfs-list-subvols" GRUB command to enumerate subvolumes on a
Btrfs device. Support output filtering and formatting with options for
path-only, id-only, read-only, and storing output in an environment
variable.

Signed-off-by: Michael Chang <mchang@suse.com>
Reviewed-by: Neal Gompa <neal@gompa.dev>
Reviewed-by: Leo Sandoval <lsandova@redhat.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/160>
---
 grub-core/fs/btrfs.c | 412 +++++++++++++++++++++++++++++++++++++++++++
 include/grub/btrfs.h |   6 +-
 2 files changed, 417 insertions(+), 1 deletion(-)

diff --git a/grub-core/fs/btrfs.c b/grub-core/fs/btrfs.c
index 5948530ff..5714642af 100644
--- a/grub-core/fs/btrfs.c
+++ b/grub-core/fs/btrfs.c
@@ -42,6 +42,8 @@
 #include <grub/diskfilter.h>
 #include <grub/safemath.h>
 #include <grub/command.h>
+#include <grub/env.h>
+#include <grub/extcmd.h>
 
 GRUB_MOD_LICENSE ("GPLv3+");
 
@@ -194,6 +196,14 @@ struct grub_btrfs_leaf_descriptor
   } *data;
 };
 
+struct grub_btrfs_root_ref
+{
+  grub_uint64_t dirid;
+  grub_uint64_t sequence;
+  grub_uint16_t name_len;
+  const char name[0];
+} GRUB_PACKED;
+
 struct grub_btrfs_time
 {
   grub_int64_t sec;
@@ -239,6 +249,10 @@ struct grub_btrfs_extent_data
 
 #define GRUB_BTRFS_OBJECT_ID_CHUNK 0x100
 
+#define GRUB_BTRFS_FS_TREE_OBJECTID 5ULL
+#define GRUB_BTRFS_ROOT_REF_KEY     156
+#define GRUB_BTRFS_ROOT_ITEM_KEY     132
+
 static grub_disk_addr_t superblock_sectors[] = { 64 * 2, 64 * 1024 * 2,
   256 * 1048576 * 2, 1048576ULL * 1048576ULL * 2
 };
@@ -248,6 +262,13 @@ grub_btrfs_read_logical (struct grub_btrfs_data *data,
 			 grub_disk_addr_t addr, void *buf, grub_size_t size,
 			 int recursion_depth);
 
+static grub_err_t
+grub_btrfs_get_parent_subvol_path (struct grub_btrfs_data *data,
+		grub_uint64_t child_id,
+		const char *child_path,
+		grub_uint64_t *parent_id,
+		char **path_out);
+
 static grub_err_t
 read_sblock (grub_disk_t disk, struct grub_btrfs_superblock *sb)
 {
@@ -1790,6 +1811,103 @@ get_root (struct grub_btrfs_data *data, struct grub_btrfs_key *key,
   return GRUB_ERR_NONE;
 }
 
+static grub_err_t
+find_pathname(struct grub_btrfs_data *data, grub_uint64_t objectid,
+              grub_uint64_t fs_root, const char *name, char **pathname)
+{
+  grub_err_t err;
+  struct grub_btrfs_key key = {
+    .object_id = grub_cpu_to_le64 (objectid),
+    .type = GRUB_BTRFS_ITEM_TYPE_INODE_REF,
+    .offset = 0ULL,
+  };
+  struct grub_btrfs_key key_out;
+  struct grub_btrfs_leaf_descriptor desc;
+  char *p = grub_strdup (name);
+  grub_disk_addr_t elemaddr;
+  grub_size_t elemsize;
+  grub_size_t alloc = grub_strlen(name) + 1;
+
+  err = lower_bound(data, &key, &key_out, fs_root,
+                    &elemaddr, &elemsize, &desc, 0);
+  if (err)
+    {
+      free_iterator(&desc);
+      return grub_error(err, "lower_bound caught %d\n", err);
+    }
+
+  if (key_out.type != GRUB_BTRFS_ITEM_TYPE_INODE_REF)
+    next(data, &desc, &elemaddr, &elemsize, &key_out);
+
+  free_iterator(&desc);
+  if (key_out.type != GRUB_BTRFS_ITEM_TYPE_INODE_REF)
+    {
+      return grub_error(GRUB_ERR_FILE_NOT_FOUND,
+                        "Can't find inode ref for {%"PRIuGRUB_UINT64_T
+                        ", %u, %"PRIuGRUB_UINT64_T"} %"PRIuGRUB_UINT64_T
+                        "/%"PRIuGRUB_SIZE"\n",
+                        key_out.object_id, key_out.type,
+                        key_out.offset, elemaddr, elemsize);
+    }
+
+  while (key_out.type == GRUB_BTRFS_ITEM_TYPE_INODE_REF &&
+         key_out.object_id != key_out.offset)
+    {
+      struct grub_btrfs_inode_ref *inode_ref;
+      char *new;
+      grub_size_t sz;
+
+      if (grub_add (elemsize, 1, &sz))
+	return grub_error (GRUB_ERR_OUT_OF_RANGE, N_("overflow is detected"));
+
+      inode_ref = grub_malloc(sz);
+      if (!inode_ref)
+	return grub_error(GRUB_ERR_OUT_OF_MEMORY,
+			  "couldn't allocate memory for inode_ref (%"PRIuGRUB_SIZE")\n", elemsize);
+
+      err = grub_btrfs_read_logical(data, elemaddr, inode_ref, elemsize, 0);
+      if (err)
+	return grub_error(err, "read_logical caught %d\n", err);
+
+      if (grub_add (grub_le_to_cpu16 (inode_ref->n), 2, &sz) ||
+	  grub_add (alloc, sz, &alloc))
+	return grub_error (GRUB_ERR_OUT_OF_RANGE, N_("overflow is detected"));
+
+      new = grub_malloc(alloc);
+      if (!new)
+	return grub_error(GRUB_ERR_OUT_OF_MEMORY,
+			  "couldn't allocate memory for name (%"PRIuGRUB_SIZE")\n", alloc);
+
+      grub_memcpy(new, inode_ref->name, grub_le_to_cpu16 (inode_ref->n));
+      if (p)
+	{
+	  new[grub_le_to_cpu16 (inode_ref->n)] = '/';
+	  grub_strcpy (new + grub_le_to_cpu16 (inode_ref->n) + 1, p);
+	  grub_free(p);
+	}
+      else
+	new[grub_le_to_cpu16 (inode_ref->n)] = 0;
+      grub_free(inode_ref);
+
+      p = new;
+      key.object_id = key_out.offset;
+      err = lower_bound(data, &key, &key_out, fs_root, &elemaddr,
+			&elemsize, &desc, 0);
+      if (err)
+	{
+	  free_iterator(&desc);
+	  return grub_error(err, "lower_bound caught %d\n", err);
+	}
+
+      if (key_out.type != GRUB_BTRFS_ITEM_TYPE_INODE_REF)
+	next(data, &desc, &elemaddr, &elemsize, &key_out);
+      free_iterator(&desc);
+    }
+
+  *pathname = p;
+  return 0;
+}
+
 static grub_err_t
 find_path (struct grub_btrfs_data *data,
 	   const char *path, struct grub_btrfs_key *key,
@@ -2505,6 +2623,293 @@ grub_cmd_btrfs_info (grub_command_t cmd __attribute__ ((unused)), int argc,
   return GRUB_ERR_NONE;
 }
 
+static grub_err_t
+get_fs_root(struct grub_btrfs_data *data, grub_uint64_t tree,
+            grub_uint64_t objectid, grub_uint64_t offset,
+            grub_uint64_t *fs_root, grub_uint64_t *flags)
+{
+  grub_err_t err;
+  struct grub_btrfs_key key_in = {
+    .object_id = grub_cpu_to_le64 (objectid),
+    .type = GRUB_BTRFS_ROOT_ITEM_KEY,
+    .offset = grub_cpu_to_le64 (offset),
+  }, key_out;
+  grub_disk_addr_t elemaddr;
+  grub_size_t elemsize;
+  struct grub_btrfs_root_item ri;
+
+  err = lower_bound(data, &key_in, &key_out, tree,
+                    &elemaddr, &elemsize, NULL, 0);
+  if (err)
+    return err;
+
+  if (key_out.type != GRUB_BTRFS_ITEM_TYPE_ROOT_ITEM || elemaddr == 0)
+    {
+      return grub_error(GRUB_ERR_FILE_NOT_FOUND,
+                    N_("can't find fs root for subvol %"PRIuGRUB_UINT64_T"\n"),
+                    key_in.object_id);
+    }
+
+  err = grub_btrfs_read_logical (data, elemaddr, &ri, sizeof (ri), 0);
+  if (err)
+    return err;
+
+  *fs_root = ri.tree;
+
+  if (flags)
+    *flags = grub_le_to_cpu64 (ri.flags);
+
+  return GRUB_ERR_NONE;
+}
+
+static const struct grub_arg_option opt_list_subvols[] = {
+  {"output", 'o', 0, N_("Output to a variable instead of the console."),
+   N_("VARNAME"), ARG_TYPE_STRING},
+  {"path-only", 'p', 0, N_("Show only the path of the subvolume."), 0, 0},
+  {"id-only", 'n', 0, N_("Show only the id of the subvolume."), 0, 0},
+  {"read-only", 'r', 0, N_("Show only read-only subvolume."), 0, 0},
+  {0, 0, 0, 0, 0, 0}
+};
+
+static grub_err_t
+grub_cmd_btrfs_list_subvols (struct grub_extcmd_context *ctxt,
+			     int argc, char **argv)
+{
+  struct grub_btrfs_data *data = NULL;
+  grub_device_t dev = NULL;
+  char *devname;
+  grub_uint64_t tree;
+  struct grub_btrfs_key key_in = {
+    .object_id = grub_cpu_to_le64_compile_time (GRUB_BTRFS_FS_TREE_OBJECTID),
+    .type = GRUB_BTRFS_ROOT_REF_KEY,
+    .offset = 0ULL,
+  }, key_out;
+  struct grub_btrfs_leaf_descriptor desc = {0};
+  grub_disk_addr_t elemaddr;
+  grub_uint64_t fs_root = 0;
+  grub_size_t elemsize;
+  int r = 0;
+  grub_err_t err;
+  int print = 1;
+  int path_only = ctxt->state[1].set;
+  int num_only = ctxt->state[2].set;
+  int read_only = ctxt->state[3].set;
+  char *varname = NULL;
+  char *output = NULL;
+
+  if (ctxt->state[0].set) {
+    varname = ctxt->state[0].arg;
+    print = 0;
+  }
+
+  if (argc < 1)
+    return grub_error (GRUB_ERR_BAD_ARGUMENT, "device name required");
+
+  devname = grub_file_get_device_name(argv[0]);
+  if (!devname)
+    return grub_errno;
+
+  dev = grub_device_open (devname);
+  grub_free (devname);
+  if (!dev)
+    return grub_errno;
+
+  data = grub_btrfs_mount(dev);
+  if (!data)
+    {
+      grub_device_close (dev);
+      return grub_error (GRUB_ERR_BAD_ARGUMENT, "could not open device");
+    }
+
+  tree = data->sblock.root_tree;
+  err = get_fs_root(data, tree, GRUB_BTRFS_FS_TREE_OBJECTID,
+                    0ULL, &fs_root, NULL);
+  if (err)
+    goto out;
+
+  err = lower_bound(data, &key_in, &key_out, tree,
+                    &elemaddr, &elemsize, &desc, 0);
+
+  if (err)
+    goto out;
+
+  do
+    {
+      char *p = NULL;
+      grub_uint64_t id = grub_le_to_cpu64 (key_out.offset);
+      grub_uint64_t flags;
+
+      if (key_out.type != GRUB_BTRFS_ITEM_TYPE_ROOT_REF || elemaddr == 0)
+        continue;
+
+      err = get_fs_root (data, tree, id, (grub_uint64_t) -1, &fs_root, &flags);
+      if (err != GRUB_ERR_NONE)
+        {
+          grub_print_error ();
+          continue;
+        }
+
+      if (read_only && !(flags & GRUB_BTRFS_ROOT_SUBVOL_RDONLY))
+        continue;
+
+      if (id == GRUB_BTRFS_ROOT_VOL_OBJECTID)
+        p = grub_strdup ("");
+      else
+        while (id != GRUB_BTRFS_ROOT_VOL_OBJECTID)
+          {
+            grub_uint64_t parent_id;
+            char *path_out;
+
+            err = grub_btrfs_get_parent_subvol_path (data, id, p, &parent_id, &path_out);
+            grub_free (p);
+            if (err)
+              {
+                grub_print_error ();
+                goto out;
+              }
+            p = path_out;
+            id = parent_id;
+          }
+
+      if (print)
+        {
+          if (num_only)
+            grub_printf("ID %"PRIuGRUB_UINT64_T"\n", grub_le_to_cpu64 (key_out.offset));
+          else if (path_only)
+            grub_printf("%s\n", p);
+          else
+            grub_printf("ID %"PRIuGRUB_UINT64_T" path %s\n", grub_le_to_cpu64 (key_out.offset), p);
+        }
+      else
+	{
+          char *old = output;
+          if (num_only)
+            output = grub_xasprintf("%s%"PRIuGRUB_UINT64_T"\n",
+                                    old ?: "", grub_le_to_cpu64 (key_out.offset));
+          else if (path_only)
+            output = grub_xasprintf("%s%s\n", old ?: "", p);
+          else
+            output = grub_xasprintf("%sID %"PRIuGRUB_UINT64_T" path %s\n",
+                                    old ?: "", grub_le_to_cpu64 (key_out.offset), p);
+          if (old)
+            grub_free(old);
+        }
+
+      grub_free (p);
+
+    } while ((r = next(data, &desc, &elemaddr, &elemsize, &key_out)) > 0);
+
+  if (output)
+    grub_env_set(varname, output);
+
+out:
+  grub_free (output);
+  free_iterator(&desc);
+
+  if (data)
+    grub_btrfs_unmount(data);
+
+  if (dev)
+    grub_device_close (dev);
+
+  return GRUB_ERR_NONE;
+}
+
+static grub_err_t
+grub_btrfs_get_parent_subvol_path (struct grub_btrfs_data *data,
+		grub_uint64_t child_id,
+		const char *child_path,
+		grub_uint64_t *parent_id,
+		char **path_out)
+{
+  grub_uint64_t fs_root = 0;
+  struct grub_btrfs_key key_in = {
+    .object_id = grub_cpu_to_le64 (child_id),
+    .type = GRUB_BTRFS_ITEM_TYPE_ROOT_BACKREF,
+    .offset = 0ULL,
+  }, key_out;
+  struct grub_btrfs_root_ref *ref;
+  char *buf;
+  struct grub_btrfs_leaf_descriptor desc;
+  grub_size_t elemsize;
+  grub_disk_addr_t elemaddr;
+  grub_err_t err;
+  char *parent_path = NULL;
+  grub_size_t sz;
+
+  *parent_id = 0;
+  *path_out = 0;
+
+  err = lower_bound(data, &key_in, &key_out, data->sblock.root_tree,
+                    &elemaddr, &elemsize, &desc, 0);
+  if (err)
+    return err;
+
+  if (key_out.type != GRUB_BTRFS_ITEM_TYPE_ROOT_BACKREF || elemaddr == 0)
+    next(data, &desc, &elemaddr, &elemsize, &key_out);
+
+  if (key_out.type != GRUB_BTRFS_ITEM_TYPE_ROOT_BACKREF)
+    {
+      free_iterator(&desc);
+      return grub_error(GRUB_ERR_FILE_NOT_FOUND, N_("can't find root backrefs"));
+    }
+
+  if (grub_add (elemsize, 1, &sz))
+    {
+      free_iterator(&desc);
+      return grub_error (GRUB_ERR_OUT_OF_RANGE, N_("overflow is detected"));
+    }
+
+  buf = grub_malloc(sz);
+  if (!buf)
+    {
+      free_iterator(&desc);
+      return grub_errno;
+    }
+
+  err = grub_btrfs_read_logical(data, elemaddr, buf, elemsize, 0);
+  if (err)
+    {
+      grub_free(buf);
+      free_iterator(&desc);
+      return err;
+    }
+
+  buf[elemsize] = 0;
+  ref = (struct grub_btrfs_root_ref *)buf;
+
+  err = get_fs_root(data, data->sblock.root_tree, grub_le_to_cpu64 (key_out.offset),
+                    (grub_uint64_t) -1, &fs_root, NULL);
+  if (err)
+    {
+      grub_free(buf);
+      free_iterator(&desc);
+      return err;
+    }
+
+  err = find_pathname(data, grub_le_to_cpu64 (ref->dirid), fs_root, ref->name, &parent_path);
+  if (err)
+    {
+      grub_free(buf);
+      free_iterator(&desc);
+      return err;
+    }
+
+  if (child_path)
+    {
+      *path_out = grub_xasprintf ("%s/%s", parent_path, child_path);
+      grub_free (parent_path);
+    }
+  else
+    *path_out = parent_path;
+
+  *parent_id = grub_le_to_cpu64 (key_out.offset);
+
+  grub_free(buf);
+  free_iterator(&desc);
+  return GRUB_ERR_NONE;
+}
+
 static struct grub_fs grub_btrfs_fs = {
   .name = "btrfs",
   .fs_dir = grub_btrfs_dir,
@@ -2521,6 +2926,7 @@ static struct grub_fs grub_btrfs_fs = {
 };
 
 static grub_command_t cmd_info;
+static grub_extcmd_t cmd_list_subvols;
 
 GRUB_MOD_INIT (btrfs)
 {
@@ -2530,10 +2936,16 @@ GRUB_MOD_INIT (btrfs)
 				    grub_cmd_btrfs_info,
 				    "DEVICE",
 				    "Print BtrFS info about DEVICE.");
+  cmd_list_subvols = grub_register_extcmd("btrfs-list-subvols",
+					 grub_cmd_btrfs_list_subvols, 0,
+					 "[-p|-n] [-r] [-o var] DEVICE",
+					 "Print list of BtrFS subvolumes on "
+					 "DEVICE.", opt_list_subvols);
 }
 
 GRUB_MOD_FINI (btrfs)
 {
   grub_fs_unregister (&grub_btrfs_fs);
   grub_unregister_command (cmd_info);
+  grub_unregister_extcmd (cmd_list_subvols);
 }
diff --git a/include/grub/btrfs.h b/include/grub/btrfs.h
index 9aad88754..19de49311 100644
--- a/include/grub/btrfs.h
+++ b/include/grub/btrfs.h
@@ -29,6 +29,7 @@ enum
     GRUB_BTRFS_ITEM_TYPE_ROOT_ITEM = 0x84,
     GRUB_BTRFS_ITEM_TYPE_ROOT_BACKREF = 0x90,
     GRUB_BTRFS_ITEM_TYPE_DEVICE = 0xd8,
+    GRUB_BTRFS_ITEM_TYPE_ROOT_REF = 0x9c,
     GRUB_BTRFS_ITEM_TYPE_CHUNK = 0xe4
   };
 
@@ -38,11 +39,14 @@ enum
     GRUB_BTRFS_TREE_ROOT_OBJECTID = 0x100,
   };
 
+#define GRUB_BTRFS_ROOT_SUBVOL_RDONLY	(1ULL << 0)
+
 struct grub_btrfs_root_item
 {
   grub_uint8_t dummy[0xb0];
   grub_uint64_t tree;
-  grub_uint64_t inode;
+  grub_uint8_t dummy2[0x18];
+  grub_uint64_t flags;
 } GRUB_PACKED;
 
 struct grub_btrfs_key
-- 
2.55.0

